File: android/src/main/java/com/hemangkumar/capacitorgooglemaps/MapPreferencesAppearance.java
package com.hemangkumar.capacitorgooglemaps;

import androidx.annotation.Nullable;

import com.getcapacitor.JSObject;
import com.google.android.gms.maps.model.MapStyleOptions;

public class MapPreferencesAppearance {
    public Integer type;
    public MapStyleOptions style;
    public Boolean isBuildingsShown;
    public Boolean isIndoorShown;
    public Boolean isMyLocationDotShown;
    public Boolean isTrafficShown;

    public static final String TYPE_KEY = "type";
    public static final String STYLE_KEY = "style";
    public static final String BUILDINGS_SHOWN_KEY = "isBuildingsShown";
    public static final String INDOOR_SHOWN_KEY = "isIndoorShown";
    public static final String MY_LOCATION_DOT_SHOWN_KEY = "isMyLocationDotShown";
    public static final String TRAFFIC_SHOWN_KEY = "isTrafficShown";

    public MapPreferencesAppearance() {
        this.type = 1;
        this.style = null;
        this.isBuildingsShown = true;
        this.isIndoorShown = true;
        this.isMyLocationDotShown = false;
        this.isTrafficShown = false;
    }

    public void updateFromJSObject(@Nullable JSObject jsObject) {
        if (jsObject != null) {
            // update mapType
            if (jsObject.has(TYPE_KEY)) {
                Integer mapType = jsObject.getInteger(TYPE_KEY, 1);
                if (mapType != null) {
                    if (mapType < 0 || mapType > 4) {
                        mapType = 1;
                    }
                    this.type = mapType;
                }
            }
            if (jsObject.has(STYLE_KEY)) {
                String mapStyle = jsObject.getString(STYLE_KEY, null);
                if (mapStyle == null) {
                    this.style = null;
                } else {
                    this.style = new MapStyleOptions(mapStyle);
                }
            }
            if (jsObject.has(BUILDINGS_SHOWN_KEY)) {
                this.isBuildingsShown = jsObject.getBool(BUILDINGS_SHOWN_KEY);
            }
            if (jsObject.has(INDOOR_SHOWN_KEY)) {
                this.isIndoorShown = jsObject.getBool(INDOOR_SHOWN_KEY);
            }
            if (jsObject.has(MY_LOCATION_DOT_SHOWN_KEY)) {
                this.isMyLocationDotShown = jsObject.getBool(MY_LOCATION_DOT_SHOWN_KEY);
            }
            if (jsObject.has(TRAFFIC_SHOWN_KEY)) {
                this.isTrafficShown = jsObject.getBool(TRAFFIC_SHOWN_KEY);
            }
        }
    }
}

File: ios/Sources/MapPreferencesAppearance.swift
import Capacitor
import GoogleMaps

class MapPreferencesAppearance {
    public static let TYPE_KEY: String! = "type"
    public static let STYLE_KEY: String! = "style"
    public static let BUILDINGS_SHOWN_KEY: String! = "isBuildingsShown"
    public static let INDOOR_SHOWN_KEY: String! = "isIndoorShown"
    public static let MY_LOCATION_DOT_SHOWN_KEY: String! = "isMyLocationDotShown"
    public static let TRAFFIC_SHOWN_KEY: String! = "isTrafficShown"

    var type: GMSMapViewType = GMSMapViewType.normal;

    var style: GMSMapStyle? = nil;

    var _isBuildingsShown: Bool = true
    var isBuildingsShown: Bool! {
        get {
            return _isBuildingsShown
        }
        set (newVal) {
            _isBuildingsShown = newVal ?? true
        }
    }

    var _isIndoorShown: Bool = true
    var isIndoorShown: Bool! {
        get {
            return _isIndoorShown
        }
        set (newVal) {
            _isIndoorShown = newVal ?? true
        }
    }

    var _isMyLocationDotShown: Bool = false
    var isMyLocationDotShown: Bool! {
        get {
            return _isMyLocationDotShown
        }
        set (newVal) {
            _isMyLocationDotShown = newVal ?? false
        }
    }

    var _isTrafficShown: Bool = false
    var isTrafficShown: Bool! {
        get {
            return _isTrafficShown
        }
        set (newVal) {
            _isTrafficShown = newVal ?? false
        }
    }

    func updateFromJSObject(object: JSObject) {
        if (object[MapPreferencesAppearance.TYPE_KEY] != nil) {
            let type = object[MapPreferencesAppearance.TYPE_KEY] as? Int;
            if (type != nil) {
                if (type == 0) {
                    self.type = GMSMapViewType.none;
                } else if (type == 1) {
                    self.type = GMSMapViewType.normal;
                } else if (type == 2) {
                    self.type = GMSMapViewType.satellite;
                } else if (type == 3) {
                    self.type = GMSMapViewType.terrain;
                } else if (type == 4) {
                    self.type = GMSMapViewType.hybrid;
                } else {
                    self.type = GMSMapViewType.normal;
                }
            }
        }
        
        if (object[MapPreferencesAppearance.STYLE_KEY] != nil) {
            let style = object[MapPreferencesAppearance.STYLE_KEY] as? String;
            if (style == nil) {
                self.style = nil;
            } else {
                do {
                    try self.style = GMSMapStyle(jsonString: style ?? "{}");
                } catch {
                    print("error parsing style json string");
                }
            }
        }
        
        self.isBuildingsShown = object[MapPreferencesAppearance.BUILDINGS_SHOWN_KEY] as? Bool;
        self.isIndoorShown = object[MapPreferencesAppearance.INDOOR_SHOWN_KEY] as? Bool;
        self.isMyLocationDotShown = object[MapPreferencesAppearance.MY_LOCATION_DOT_SHOWN_KEY] as? Bool;
        self.isTrafficShown = object[MapPreferencesAppearance.TRAFFIC_SHOWN_KEY] as? Bool;
    }
    
    func getJSObject(_ mapView: GMSMapView) -> JSObject {
        var type = 1;
        if (mapView.mapType == GMSMapViewType.none) {
            type = 0;
        } else if (mapView.mapType == GMSMapViewType.normal) {
            type = 1;
        } else if (mapView.mapType == GMSMapViewType.satellite) {
            type = 2;
        } else if (mapView.mapType == GMSMapViewType.terrain) {
            type = 3;
        } else if (mapView.mapType == GMSMapViewType.hybrid) {
            type = 4;
        }
        
        return [
            MapPreferencesAppearance.TYPE_KEY: type,
            MapPreferencesAppearance.BUILDINGS_SHOWN_KEY: mapView.isBuildingsEnabled,
            MapPreferencesAppearance.INDOOR_SHOWN_KEY: mapView.isIndoorEnabled,
            MapPreferencesAppearance.MY_LOCATION_DOT_SHOWN_KEY: mapView.isMyLocationEnabled,
            MapPreferencesAppearance.TRAFFIC_SHOWN_KEY: mapView.isTrafficEnabled
        ]
    }
}
